File: NewMacros/EventLoop.cpp
#include "EventLoop.h"
#include <iostream>
#include <stdexcept>

EventLoop::EventLoop() {
 // nothing to do here
}

void EventLoop::initialize() {
 // create an instance of the TChain class
 m_chain = new TChain(treeName);

 // loop through the input files and add them to the chain
 for(auto inputFile : inputFiles) {
   m_chain->Add(inputFile);
   std::cout << "Added file: " << inputFile << std::endl;
 }

 // create an instance of the Data class. Here the variables
 // are linked with the tree using the SetBranchAddress method
 m_data = new Data(m_chain);

  // initialise the algorithms
 for(auto algorithm : algorithms) {
  algorithm->initialize(m_data);
 }
}

void EventLoop::execute() {
 // sanity check. m_chain must not be zero
 if(!m_chain) {
   throw std::runtime_error("Calling execute while the wvent loop was not initialized.");
 }

 // here we do the actual event loop
 for(int i=0; i<m_chain->GetEntries(); ++i) {
   // event number printout
   if(i%1000==0) {
     std::cout << "Event " << i << std::endl;
   }

   // read the data for i-th event
   m_chain->GetEntry(i);


  //  // now we can work with the variables. Let's for example print electron mass 
  // // but only for every 1000th event
  // if(i%1000==0) {
  //  std::cout << "pt electron = " << m_data->patElectron_pt->at(0) << std::endl;
  //}
  // execute all the algorithms attached to this event loop
  for(auto algorithm : algorithms) {
    algorithm->execute();
  }

 }

}

                

File: NewMacros/EventLoop.py
from ROOT import EventLoop as EventLoopCpp

class EventLoop(object):
 """ Wrapper around the c++ class event loop
 """ 
 def __init__(self, name):
  # set the event loop name
  self.name = name

  # now we can create instance of the c++ class EventLoop
  self.eventLoop = EventLoopCpp()

  # set the tree name attribute
  self.eventLoop.treeName = "BOOM"

  # keep the list of python Algorithm class instances
  self.algs = []

 def addAlgorithm(self, algorithm):
  """ add an algorithm into this event loop
  """
  algorithm.setSumw2()
  self.algs += [ algorithm ]
  self.eventLoop.algorithms.push_back(algorithm.alg)
 
 def addAlgorithms(self, algorithms):
  """ add multiple algorithms into this event loop
  """
  for alg in algorithms:
   self.addAlgorithm(alg)

 def execute(self):
  """ initialize and execute the event loop
  """
  self.eventLoop.initialize()
  self.eventLoop.execute()

 def save(self):
  """ save histograms from all algorithms
  """
  for alg in self.algs:
   alg.save(self.name)